### ISSUE TITLE
[Request deployment] New Hub: NASA ESDIS Hub

### ISSUE THREAD

**Issue by colliand** [2i2c-org/infrastructure#3068]
### The GitHub handle of the community representative

@bilts

### Hub important dates

- **Target Start Date**: 2023-09-22
- **Required Start Date**: 2023-09-30

### Hub Authentication Type

GitHub (e.g., @mygithubhandle)

### First Hub Administrators

- Patrick Quinn (@bilts)

### [GitHub Auth only] How would you like to manage your users?

pending

### [GitHub Teams Auth only] Profile restriction based on team membership

pending

This hub should launch with configuration and image based on the OpenScapes hub. 

### Hub logo image URL

pending

### Hub logo website URL

https://www.earthdata.nasa.gov/esdis

### Hub user image GitHub repository

pending

### Hub user image tag and name

pending; likely same as openscapes

### Extra features you would like to enable

- [ ] Dedicated Kubernetes cluster
- [ ] Scalable Dask Cluster

### (Optional) Preferred cloud provider

None

### (Optional) Billing and Cloud account

None

### Other relevant information to the features above

This hub deployment request is part of 2i2c's ongoing `NASA VEDA` project. The ESDIS hub will be launched with configuration and image based on the OpenScapes hub. The hub should be deployed in the NASA SMD cluster where 2i2c is currently operating the NASA VEDA hub. 

I request confirmation or edits to these configuration plans from @freitagb or @bilts. 

### Tasks to deploy the hub

- [ ] 1. Deploy information filled in above
- [ ] 2. Engineer who will deploy the hub is assigned
- [ ] 3. If using GitHub Orgs/Teams Auth, Engineer is given Owner rights to the org to set this up.
- [ ] 4. Initial Hub deployment PR <link to PR>
- [ ] 5. Administrators able to log on -> Hub now in steady-state

**Comment by colliand** [comment]
Some information is required from @bilts to deploy this hub. 

- [ ] Please share a **URL pointing to the splash image** you'd like to appear on the login page for the ESDIS hub. See the [NASA VEDA hub login page](https://nasa-veda.2i2c.cloud/hub/login?next=%2Fhub%2F) for an example. 
- [ ] Following the pattern we've used with VEDA and GHG, I assume that access to the ESDIS hub will be managed by membership in a GitHub team or organization. Please share the URL for that GitHub team or org.

**Comment by damianavila** [comment]
I have assigned @sgibson91 for the deployment of this hub.

@colliand please provide any missing information so @sgibson91 can start with this deployment soon.
Also, can you confirm if the target/required dates are actually up to date? I have the feeling they might be outdated (but I could be wrong).


@sgibson91, feel free to request any information you may need to move this forward.

Thanks everyone!

**Comment by colliand** [comment]
I will update the dates. Thanks @damianavila for assigning this hub deployment to @sgibson91. I will work to gather the needed info for Sarah to complete the deployment.

**Comment by sgibson91** [comment]
Depending on when the required information is provided, and when Google resolves this incident, we may run into issues with Persistent Volumes when deploying this hub https://status.cloud.google.com/incidents/EBxyHQgEPnbM3Syag5yL

**Comment by sgibson91** [comment]
@colliand is there an update on this?

**Comment by colliand** [comment]
I don't have an update. We are waiting for direction from @freitagb and colleagues at NASA.

**Comment by damianavila** [comment]
OK, I removed it from the current iteration and we can re-add it to a future sprint when we have more information.

**Comment by sgibson91** [comment]
> We have a repository, [nasa/Earthdata-cloud-hub](https://github.com/nasa/Earthdata-cloud-hub) within the NASA GitHub and a team within the NASA org who can maintain that repo. Initially, we want members of this team to have access. I don't know how best to accomplish that in the 2i2c setup.

Can you give us the name of the GitHub Team? That's all I require from you to configure access for them

> > ### Hub user image tag and name
> 
> I don't know what this means

So there is usually a Docker image that describes a software environment that is pushed to an image registry such as quay.io. We can reference this image using a name and tag. This Docker image will be used to set the computational environment (available packages, etc) for every person who logs into the hub. There is a note in the issue body that says:

> The ESDIS hub will be launched with configuration and image based on the OpenScapes hub.

These are the images that OpenScapes uses:
- https://hub.docker.com/r/openscapes/python/tags
- https://hub.docker.com/r/openscapes/rocker/tags
- https://hub.docker.com/r/openscapes/matlab/tags

Does that ring true from your conversations with Jim and/or James @bilts?

**Comment by yuvipanda** [comment]
@sgibson91 My suggestion for the images is to copy what we have for the GHG hub, and we can modify from there later if needed. Does that sound ok, @freitagb?

**Comment by sgibson91** [comment]
I invite @jmunroe and @colliand to work with @freitagb and @bilts to establish a path forward here

**Comment by freitagb** [comment]
@yuvipanda @sgibson91 - can we take the images used in the Openscapes hub? It may also be useful to make the QGIS image available as well. I've also reached out about github org/team access. @bilts shared a repo, but i'm not sure of the associated org/team we can use to provision access.

**Comment by bilts** [comment]
I have a logo image but nowhere to host it.

**Comment by yuvipanda** [comment]
Ok, so let's just copy the profileList from https://github.com/2i2c-org/infrastructure/blob/e9a7ac7802c5bc8855767ae03c2954743ff974f4/config/clusters/openscapes/prod.values.yaml#L14 to start with, and add https://github.com/2i2c-org/infrastructure/blob/e9a7ac7802c5bc8855767ae03c2954743ff974f4/config/clusters/nasa-ghg/common.values.yaml#L100 for qgis as a profile option.

**Comment by freitagb** [comment]
> (edit) I'd prefer to copy what we have for the Openscapes hub.

Is this in response to access control or the images?

**Comment by yuvipanda** [comment]
> The hub should be deployed in the NASA SMD cluster where 2i2c is currently operating the NASA VEDA hub.

From the original description is also not right, as this will be a different AWS account (I just confirmed this with @freitagb). As I'm the only person so far who has set up clusters inside SMCE AWS accounts, I wrote up docs on how to do that in https://github.com/2i2c-org/infrastructure/pull/3229.

@bilts @freitagb assuming that this is going to be in a new AWS account, can you provision IAM user accounts for @sgibson91 and provide her with credentials? You'll also have to approve a MFA exemption for our continuous deployment pipeline once we have it - I think @freitagb is by now familiar with the process :)

**Comment by freitagb** [comment]
Thanks @bilts - i see Openscapes used CILogin for authentication @yuvipanda and @sgibson91. Can you confirm that you are recommending we use github org/team for authentication as we have with VEDA and GHG hubs?

**Comment by sgibson91** [comment]
From @colliand in Slack:

> Let's deploy the ESDIS on a new AWS cluster with billing account owned by SMCE.

---

> i see Openscapes used CILogin for authentication @yuvipanda and @sgibson91. Can you confirm that you are recommending we use github org/team for authentication as we have with VEDA and GHG hubs?

@freitagb It really is up to you, we support both methods of authentication. If you expect to have users that overlap with GHG, VEDA, and ESDIS, it might be nice for them to be presented with the same kind of login experience across all the hubs.

**Comment by freitagb** [comment]
Ultimately the decision comes back to @bilts - just trying to make sure there's clear understanding 😄

**Comment by yuvipanda** [comment]
Yes my recommendation is that we continue using GitHub authentication here, as you can't use teams with CILogon.

**Comment by colliand** [comment]
Thanks all for this discussion. I recommend @bilts specifies a GitHub org or GitHub team (by sharing the URL) that will be used to manage access. Users who belong to that Team or Org will be able to access the hub using their GitHub credentials.

**Comment by sgibson91** [comment]
I was given a login to an ESDIS AWS account last night. I'll set myself up and make sure all 2i2c engineers are added as a first step.

**Comment by sgibson91** [comment]
Ok, I have deployed a cluster into the account. @freitagb can you please begin the process of having the hub-continuous-deployer user exempted from MFA please?

**Comment by bilts** [comment]
We have a team nasa/earthdata-cloud-hub-maintainer. It's non-public. Can that be used?

**Comment by sgibson91** [comment]
Yes, I think that can be used. Is it possible to give me admin rights to the org temporarily so I can properly set up the OAuth application? This is our standard operating procedure as documented here https://infrastructure.2i2c.org/hub-deployment-guide/configure-auth/github-orgs/#get-admin-access-to-the-target-organisation and I will remove myself from the org once I have confirmed authentication is working as expected.

**Comment by sgibson91** [comment]
I can see from the AWS console that the hub-continuous-deployer user is no longer in the SMCE-UserRestrictions group. I should be unblocked, but the deployer is not yet working as I'd expect it to. I'm investigating.

**Comment by freitagb** [comment]
@sgibson91 - I've received notification that the MFA exception has been applied. Can you confirm everything works as expected?

**Comment by sgibson91** [comment]
Hi @freitagb, yes everything's working now. I'll get the hubs deployed. Is there any update on getting me access to the GitHub org so I properly configure auth? https://github.com/2i2c-org/infrastructure/issues/3068#issuecomment-1756195481

**Comment by sgibson91** [comment]
> I have a logo image but nowhere to host it.

@bilts are you able upload it to something like [imgur](https://imgur.com/)

Also, do you have a link for the org providing the funding for this hub?

**Comment by yuvipanda** [comment]
I think imgur is unfortunately far too along in its enshittification for us to use it. Instead, let's make a github repo and use direct link from *that*?

**Comment by freitagb** [comment]
It seems like @bilts is on international travel this week so i'll do my best to address your questions. I've asked for access to the github org/team, but perhaps you can use ` veda-analytics-access/veda-ghg-admins` even though i believe auth has already been set up on that org. Let me know if you would like access to that team and i'll add you. As for the url, i think that's fine to start and if we need to change it in the future, we can do that at a later date. Does that work?

**Comment by freitagb** [comment]
Hi @sgibson91 - I spoke with @afriesz and @cpkrehbiel and access to the `nasa/earthdata-cloud-hub-maintainer` requires a NASA account. Could we work together to configure auth on the org/team?

**Comment by sgibson91** [comment]
The URL for the prod hub is https://esdis.2i2c.cloud

You will need to go through the same process as above when you authenticate for the first time

**Comment by freitagb** [comment]
Hi @sgibson91 - i forgot the *.2i2c.cloud domains are blocked by default and require an exemption. @bilts can decide if we should do the exception or provide a different url. For now we can leave it as is. With that said, if we want to test hub functionality before access is available to the `nasa/earthdata-cloud-hub-maintainer` would it be possible to add the `veda-analytics-access/esdis-hub-beta-testers` team to the allowed organization list?

**Comment by sgibson91** [comment]
@freitagb done!

**Comment by freitagb** [comment]
Thanks! I also was looking to make an edit to the org/team @bilts provided based on feedback from @amfriesz. Can we update the `earthdata-cloud-hub-maintainer` with `earthdata-cloud-playground-maintainer`? We still need guidance on the target url from @bilts, but that should be the last change. Using the other org we were able to confirm hub access and existing DAAC notebooks execute as expected.

**Comment by sgibson91** [comment]
Done!

**Comment by amfriesz** [comment]
@sgibson91, I am a member of the `earthdata-cloud-playground-maintainer` GitHub team, but I'm unable to access the ESDIS hub ([https://esdis.2i2c.cloud/](https://esdis.2i2c.cloud/hub/login)). I was able to access the hub when @brianmfreitag added me to `esdis-hub-beta-testers`

**Comment by freitagb** [comment]
I think this an issue with setting up auth on the org/team. @sgibson91, can we add  `nasa:lpdaac_science` to allow @amfriesz and @cpkrehbiel to test the authentication process in their own repo?

**Comment by amfriesz** [comment]
`nasa:lpdaac_science` is within the NASA GitHub org. I wonder is the NASA org is preventing access for those teams. I created a team in the Openscapes space. Could we try that one? https://github.com/orgs/NASA-Openscapes/teams/esdis_hub_testers

**Comment by sgibson91** [comment]
> can we add nasa:lpdaac_science to allow @amfriesz and @cpkrehbiel to test the authentication process in their own repo?

@freitagb @amfriesz I can of course add this team, but the authentication happens at the _organisation_ level (so `nasa`), **not** the team level (not `lpdaac_science` or whatever repo that team uses). So @amfriesz you are exactly right. I will add the NASA-Openscapes/esdis_hub_testers team

When either of you logged in, were you presented with a screenshot similar to the one below?

![image](https://github.com/2i2c-org/infrastructure/assets/44771837/20427711-b265-49e1-a490-cd21d129f06d)

If so, the "Approve" or "Request" button next to the NASA org needed to be clicked **before** clicking the green "Authorize 2i2c-org" button at the bottom. If it said "Request" next to the NASA org, I suspect we are waiting on the admins of that org to approve the application. Until they do that, I don't think anyone associated with a team located in the NASA org will be allowed to log in, and there's nothing I can do from my end other than revoke all user tokens generated by the OAuth app and force everyone to see the above screenshot again. This is why we (2i2c engineers) ask for temporary Owner rights on the organisation, to prevent this kind of back-and-forth.

**Comment by freitagb** [comment]
yes. Thank you @sgibson91 - i think this is making a few of us at least consider how the official nasa github is incorporated here if at all. We can revisit once we get more guidance on that.

**Comment by damianavila** [comment]
Hey folks, there are yet some missing pieces on this one such as:

* The URL you want to use (apparently .2i2c.cloud domain gets blocked)
* The GitHub Teams you want to authenticate with (originally you were using the official NASA org, but the restrictions are so tight on it that we cannot get the OAuth apps set up correctly)
* Funding info for the homepage
* Logo for the homepage

@sgibson91, if there is something else missing (or something I got wrong, please correct me).

cc @freitagb @bilts @amfriesz who are the people involved so far from the community side.

cc @colliand @jmunroe, who might need to ping the hub champions to get these pieces.

**Comment by damianavila** [comment]
> cc @colliand @jmunroe, who might need to ping the hub champions to get these pieces.

@colliand @jmunroe, we need to find a closure to this deployment issue, can we make sure we get the information to finish it ASAP? Thanks!!

**Comment by bilts** [comment]
@damianavila Thanks for your patience. We've had some issues on our end.

* The esdis.2i2c.cloud domain can now be used within our firewall, so we can keep that.
* I created a team here: https://github.com/orgs/nasa-esdis/teams/cloud-users … it’s still ideal if we can use a NASA one, though
* Funding info... "Funded by: [NASA](https://www.earthdata.nasa.gov/esds)" is sufficient
* Logo... below

<img width="713" alt="Screenshot 2023-11-28 at 2 08 05 PM" src="https://github.com/2i2c-org/infrastructure/assets/61120/3380676a-1f2e-400d-8471-79496510c1e7">

**Comment by sgibson91** [comment]
Thank you @bilts!

> * I created a team here: [github.com/orgs/nasa-esdis/teams/cloud-users](https://github.com/orgs/nasa-esdis/teams/cloud-users)

Ok, I have added this team and _also_ removed all the other teams that were added while we were debugging. Let me know if I should add any of them back.

> … it’s still ideal if we can use a NASA one, though

Unfortunately there is nothing we at 2i2c can do about that. We need the owners of the NASA org to either grant us owner rights on the org to setup our authentication ourselves (requires getting a NASA email address which is probably not worth it), or they have to approve the OAuth application themselves. If you make any progress with that and want to change the team in the future, we can absolutely do that!

---

Hubs are available at:

- https://staging.esdis.2i2c.cloud
- https://esdis.2i2c.cloud
